[api-minor] Ensure that the `getDocument` Promise is rejected if the `loadingTask` is destroyed, or an `Error` is thrown, inside of the `onPassword` callback (issue 7806)

Snuffleupagus opened the pull request:
Tentatively fixes #7806 (and replaces PR #7924).

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/mozilla/pdf.js/7926)
<!-- Reviewable:end -->

Snuffleupagus committed:
[api-minor] Ensure that the `getDocument` Promise is rejected if the `loadingTask` is destroyed, or an `Error` is thrown, inside of the `onPassword` callback (issue 7806)

This patch also removes the `UpdatePassword` message, in favour of using the `sendWithPromise` method of `MessageHandler`.
Furthermore, the patch also refactors the `BasePdfManager_updatePassword`/`BasePdfManager_passwordChanged` methods (in pdf_manager.js), and the `pdfManagerReady` function (in worker.js).

yurydelendik commented on src/display/api.js:
```diff
@@ -1508,28 +1512,28 @@ var WorkerTransport = (function WorkerTransportClosure() {
+        var error = new PasswordException(exception.message, exception.code);
+        loadingTask._capability.reject(error);
+
+        this._passwordCapability.reject(error);
```
can you check for `!loadingTask.onPassword` and returned rejected promise (with PasswordException).

yurydelendik commented on src/display/api.js:
```diff
@@ -1508,28 +1512,28 @@ var WorkerTransport = (function WorkerTransportClosure() {
+        return this._passwordCapability.promise;
       }, this);
 
+
```
nit: remove unneeded line

yurydelendik commented on src/core/worker.js:
```diff
@@ -671,11 +671,9 @@ var WorkerMessageHandler = {
-          }
+          handler.sendWithPromise('Password', e).then(function (data) {
+            pdfManager.updatePassword(data.password);
+          });
```
add `.catch(function (e) { handler.send('PasswordException', e); });`. (I'm not sure if we need to re-wrap `e`, but I think this will work out) We need to add 'PasswordException' handler, to abort loading task. `Password` can be `PasswordRequest` to avoid confusion.

yurydelendik replied:
The code at pdfManagerReady below looks non-symmetrical to recovery password handling. I think we shall remove pdfManager.passwordChanged/._passwordChangedCapability? The pdfManagerReady call will be called after password update (we need to move pdfManagerReady from being anonymous function to closure before getPdfManager() call).

Snuffleupagus replied:
> add .catch(function (e) { handler.send('PasswordException', e); });. (I'm not sure if we need to re-wrap e, but I think this will work out)

I complication here is that the exception that we will thus `catch`, is actually a `string` rather than an actual `Error` here since `postMessage` has issues with serializing those.

For now, the patch is still sending the original `PasswordException` in this case, since I'm having trouble getting this to work properly otherwise.

yurydelendik reviewed (commented):
It getting really complicated, thanks for working on this issue.

Snuffleupagus commented on src/display/api.js:
```diff
@@ -1435,6 +1436,13 @@ var WorkerTransport = (function WorkerTransportClosure() {
 
+      if (this._passwordCapability) {
+        var err = new Error('Worker was destroyed during onPassword callback');
+        this.loadingTask._capability.reject(err);
```
*This is definitely getting more complicated than I imagined; thanks for your in-depth review comments thus far!*

I tried removing that, but it appears to be causing some problems:
When calling `loadingTask.destroy();` in the `onPassword` handler, it seems that the worker might be terminated *before* it has had a chance to send the (new) `PasswordException` (as outlined in comment https://github.com/mozilla/pdf.js/pull/7926/files#r94335620).

The `loadingTask` will thus remain in a pending state, without ever getting either resolved *or* rejected. Note that I'm seeing this both in manual tests in the viewer, as well as in the unit-tests.

How would you propose that we handle this case?
Could/should we use the `WorkerTask` (in worker.js) to track the password promise there, to ensure that worker termination waits for the rejected `_passwordCapability` from api.js?

Snuffleupagus committed:
Refactor the `password` handling so that it's stored in the `PdfManager`s, instead of in the `XRef`

We're already passing in a, currently unused, `PdfManager` instance when initializing the `XRef`. To avoid having to pass a single `password` parameter around, we could thus simply get the `password` through the `PdfManager` instance instead.

Snuffleupagus commented:
@yurydelendik Thanks for the in-depth review comments! I've tried to address them, but there's a couple of things I'm entirely sure how to handle, please see https://github.com/mozilla/pdf.js/pull/7926#discussion_r94344712 and https://github.com/mozilla/pdf.js/pull/7926#discussion_r94398272.

yurydelendik reviewed (approved):
Looks good to me. From the comments above I did not get if there are any issues, but that's what I had in mind (except WorkerTask, but it's okay to have it). Thank you for the patch.

Snuffleupagus commented:
/botio test

Snuffleupagus commented:
/botio-windows test
